servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java

```diff
@@ -0,0 +1,133 @@
+     *
+     * @return the {@link TrustManagerFactory} used for verifying the remote endpoint's certificate.
+     */
+    @Nullable
+    TrustManagerFactory trustManagerFactory();
+
+    /**
+     * Get the trusted certificates for verifying the remote endpoint's certificate. The input stream should
+     * contain an {@code X.509} certificate chain in {@code PEM} format.
+     * @return the trusted certificates for verifying the remote endpoint's certificate. The input stream should
+     * contain an {@code X.509} certificate chain in {@code PEM} format.
+     * <p>
+     * Each invocation of the {@link Supplier} should provide an independent instance of {@link InputStream} and the
+     * caller is responsible for invoking {@link InputStream#close()}.
+     */
+    @Nullable
+    Supplier<InputStream> trustCertChainSupplier();
+
+    /**
+     * Get the {@link KeyManagerFactory} to use for the SSL/TLS handshake.
+     *
+     * @return the {@link KeyManagerFactory} to use for the SSL/TLS handshake.
+     */
+    @Nullable
+    KeyManagerFactory keyManagerFactory();
+
+    /**
+     * Get a {@link InputStream} which provides {@code X.509} certificate chain in {@code PEM} format associated with
+     * {@link #keySupplier()}.
+     * @return the certificate chain associated with {@link #keySupplier()}.
+     * <p>
+     * Each invocation of the {@link Supplier} should provide an independent instance of {@link InputStream} and the
+     * caller is responsible for invoking {@link InputStream#close()}.
+     */
+    @Nullable
+    Supplier<InputStream> keyCertChainSupplier();
+
+    /**
+     * Get a {@link InputStream} which provides a {@code PKCS#8} private key in {@code PEM} format associated with
+     * {@link #keyCertChainSupplier()}.
+     * @return a {@link InputStream} which provides a {@code PKCS#8} private key in {@code PEM} format associated with
+     * {@link #keyCertChainSupplier()}.
+     * <p>
+     * Each invocation of the {@link Supplier} should provide an independent instance of {@link InputStream} and the
+     * caller is responsible for invoking {@link InputStream#close()}.
+     */
+    @Nullable
+    Supplier<InputStream> keySupplier();
+
+    /**
+     * Get the password required to access the key material (e.g. from {@link #keySupplier()}).
+     * @return the password required to access the key material (e.g. from {@link #keySupplier()}).
+     */
+    @Nullable
+    String keyPassword();
+
+    /**
+     * Get the TLS protocols to enable, in the order of preference.
+     *
+     * @return the TLS protocols to enable, in the order of preference.
+     * @see SSLEngine#setEnabledProtocols(String[])
+     */
+    @Nullable
+    List<String> sslProtocols();
+
+    /**
+     * Get the TLS <a href="https://tools.ietf.org/html/rfc7301">ALPN</a> protocols.
+     * @return the TLS <a href="https://tools.ietf.org/html/rfc7301">ALPN</a> protocols.
+     */
+    @Nullable
+    List<String> alpnProtocols();
```

Comment:
for HTTP there are a number of params we try to automatically infer, and ALPN is one of them. however the ALPN parameters are a function of the ssl configuration and don't have to be inferred by all protocols and this now unifies the SSL config API at the TCP level (e.g. `TcpClientConfig`, `TcpServerConfig`) to be focused on `SslConfig` instead of independent params.

Reply:
I understand it simplifies some internal API, but complicates public api and understanding of ssl config. I bet users will try to configure this list for http and will do that incorrectly (inconsistently with http protocols). Because it's additive, let's defer it until we have a valid use-case when this configuration is required. For now, it's easier and safer for users if we infer alpn protocols.

Reply:
I've updated the http code to always take the inferred value and clarified in the builder that the set value maybe ignored (depending upon the protocol).